### ISSUE TITLE
Add LossCompute Object in the Translator to handle copy_attn and avoid duplicate code

### ISSUE THREAD

**Issue by se4u** [OpenNMT/OpenNMT-py#777]
The `_run_target` method does not handle the special case of `copy attention`. For example,  the `CopyGenerator` requires the computed attention and `src_map` which it is currently not provided. Supporting this will basically require some extra conditional statements and the code will probably just require simple copying from the `_translate_batch` method and the `CopyGenerator` class. 

https://github.com/OpenNMT/OpenNMT-py/blob/e61589d4485a2811850f5894cf6931f8c91b5165/onmt/translate/Translator.py#L408

I am currently fixing this in my workspace. If someone more familiar with the innards of onmt has time to look into it that'll be great. I'll post a fix because it's important to me anyway.

**Comment by vince62s** [comment]
duplicate of #425

**Comment by vince62s** [comment]
@se4u did you read Henri's comment in #936 ?
Did you come up with a good fix ?

**Comment by se4u** [comment]
I initialized the same compute object that is used during training to compute the loss during translation. Then I just delegated loss computation to that object.  This seems to be the right thing to do because it minimizes code duplication and also because the training loss can help us debug whether we have optimization error or search error. [e.g. see slide 14 of Andrew Ng's ML Advice](http://cs229.stanford.edu/materials/ML-advice.pdf) 

```python
class Translator(object):
    def translate(self, src_path=None,
                  src_data_iter=None,
                  tgt_path=None,
                  tgt_data_iter=None,
                  src_dir=None,
                  batch_size=1,
                  attn_debug=False):
    ...
    if self.copy_attn:
         self.compute = onmt.modules.CopyGeneratorLossCompute(
                    generator=self.model.generator,
                    tgt_vocab=self.fields["tgt"].vocab,
                    force_copy=self.copy_attn_force,
                    normalize_by_length=self.copy_loss_by_seqlength,
                    compute_stats=False,
                    aggregate_loss=False,
                    label_smoothing=self.label_smoothing)

    def _run_target(self, batch, data):
        ...
        if self.copy_attn:
            loss, _ = self.compute._compute_loss(
                batch,
                **self.compute._make_shard_state(
                    batch,
                    dec_out,
                    (0, batch.tgt.size(0)),
                    attns))
       return loss
```

**Comment by vince62s** [comment]
reopening, see my comment in #862 for a better refactoring.
@bpopeters @Henry-E
